### PR TITLE
fix: with `io.ReadAll()` use `io.LimitReader()` for DoS prevention

### PR DESCRIPTION
**Pull request by gbaranski**
All occurences of `io.ReadAll()` now use `io.LimitReader()` before passing data to the former.
The limit for the reader has been set in [core.go](https://github.com/NordSecurity/nordvpn-linux/blob/a58ca400e7936c264438efa59413dbd455a85481/core/core.go#L24), after discussions the value has been set to `10MB`.

However, it is with some exceptions:
- [daemon/dns/hosts.go](https://github.com/NordSecurity/nordvpn-linux/blob/a58ca400e7936c264438efa59413dbd455a85481/daemon/dns/hosts.go#L88). Reading `/etc/hosts` should not be a subject to any attacks. As editing the file on it's own requires `root` access.
- [events/logger/logger.go NotifyRequestAPIVerbose()](https://github.com/NordSecurity/nordvpn-linux/blob/a58ca400e7936c264438efa59413dbd455a85481/events/logger/logger.go#L58). Marked as "Do not use in production builds"

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -565,3 +565,22 @@ func getData[T any](api *DefaultAPI, token string, url string) (T, error) {
 
 	return data, nil
 }
+
+const maxBytesLimit int64 = 1024 * 1024 * 10 // 10MB
+
+// re-implementation of io.ReadAll with max limit
+func MaxBytesReadAll(r io.Reader) ([]byte, error) {
+	limitedReader := &io.LimitedReader{
```
Comment (devzbysiu) [inline core/core.go:573]: There is also `io.LimitReader(r io.Reader, n int64)` which wraps reader in `LimitedReader` on it's own

File: core/core.go [hunk 2]
```diff
@@ -565,3 +565,22 @@ func getData[T any](api *DefaultAPI, token string, url string) (T, error) {
...
+		R: r,
+		N: maxBytesLimit,
+	}
+	data, err := io.ReadAll(limitedReader)
+	if err != nil {
+		return nil, err
+	}
+	if limitedReader.N == 0 {
```
Comment (devzbysiu) [inline core/core.go:581]: Isn't this off-by-one here? So you can't read 10MB of data now, because you'll have `N` set to 0 and it will fail, right? :thinking:

File: core/core.go [hunk 3]
```diff
@@ -565,3 +565,22 @@ func getData[T any](api *DefaultAPI, token string, url string) (T, error) {
...
+		return nil, fmt.Errorf("input exceeded the max limit")
```
Comment (devzbysiu) [inline core/core.go:582]: [small] Maybe let's add the limit value to the error msg?

**Review by bartoszWojciechO — APPROVED** [review]

File: core/core.go [hunk 4]
```diff
@@ -565,3 +565,22 @@ func getData[T any](api *DefaultAPI, token string, url string) (T, error) {
 
 	return data, nil
 }
+
+const maxBytesLimit int64 = 1024 * 1024 * 10 // 10MB
+
+// re-implementation of io.ReadAll with max limit
```
Comment (devzbysiu) [inline core/core.go:571]: I think doc comments in go should start with the method name

File: core/core_test.go [hunk 5]
```diff
@@ -224,3 +226,30 @@ func TestDefaultAPI_ServiceCredentials(t *testing.T) {
 		})
 	}
 }
+
+func TestMaxBytes(t *testing.T) {
```
Comment (devzbysiu) [inline core/core_test.go:230]: I would suggest to add tests for edge conditions. This would detect the off-by-one mentioned in other comment. Generally a lot of bugs happen at the edge conditions that's why it's super useful to have edge conditions tested if possible

File: core/core.go (hunk 4 above)
Comment (gbaranski) [reply core/core.go:571]: that's right, fixed

File: core/core.go (hunk 2 above)
Comment (gbaranski) [reply core/core.go:591]: that's right, there was a bug related to off-by-one, but not there.

I've created tests that prove that this is working now with input length of -1 +0 or +1 in difference to the limit

File: core/core_test.go (hunk 5 above)
Comment (gbaranski) [reply core/core_test.go:230]: done

**Review by devzbysiu — APPROVED** [review]

File: core/core_test.go [hunk 6]
```diff
@@ -224,3 +226,34 @@ func TestDefaultAPI_ServiceCredentials(t *testing.T) {
...
```
Comment (mariusSincovici) [inline core/core_test.go:230]: Please switch to table testing.

**Review by mariusSincovici — CHANGES_REQUESTED** [review]

**Review by mariusSincovici — APPROVED** [review]
